File: nicegui/elements/anywidget/anywidget.js
import { load_widget, load_css } from "widget";
import { cleanObject } from "../../static/utils/json.js";
import { set as set_helper } from "set-helper";

/**
 * Communication layer for anywidget models.
 *
 * @class
 * @classdesc
 * Comm is responsible for handling messages between the frontend model and the Python backend
 * via the emit_to_py callback. Used for synchronizing widget state and sending/receiving
 * protocol messages.
 *
 * @property {Object} model
 * @property {(eventType: string, ...args: any[]) => void} emit_to_py E.g. this.$emit
 * @property {string|null} comm_id Null until updated by the server
 */
class Comm {
  /**
   * Create a new Comm instance.
   * @param {Object} model
   * @param {(eventType: string, ...args: any[]) => void} emit_to_py E.g. this.$emit
   */
  constructor(model, emit_to_py) {
    this.model = model;
    this.emit_to_py = emit_to_py;
    this.comm_id = null;
  }

  recv_msg({msg_type, content: { data, comm_id }, metadata, buffers }) {

    if (this.comm_id != null && this.comm_id !== comm_id) {
      return;
    }

    switch (msg_type) {
      case "comm_open":
        if (this.comm_id != null) {
          break;
        }
        this.comm_id = comm_id;
        this.model.set_state(data.state);
        // TODO: Store target name, buffer paths, buffers, metadata
        break;
      case "comm_msg":
        switch (data.method) {
          case "update":
            this.model.set_state(data.state);
            break;
          case "echo_update":
            // In jupyter this is meant to send back the state in case there are multiple front ends.
            break;
          case "request_state":
            this.send_msg('update', { state: this.attributes });
            break;
          case "custom":
            this.model.emit('msg:custom', data, buffers);
            break;
        }
        break;
      default:
        console.warn('Unknown message type:', msg_type, ' for ', {msg_type, data, metadata, buffers, comm_id});
        break;
    }
  }

  send_msg(method, content, buffers) {
    this.emit_to_py('anywidget:msg', {
      msg_type: 'comm_msg',
      content: {
        comm_id: this.comm_id,
        data: { method, ...content },
      },
      buffers: buffers || [],
    });
  }
}

/**
 * Implement AFM: https://anywidget.dev/en/afm/
 * References:
 * - Marimo AFM impl:
 * https://github.com/marimo-team/marimo/blob/7f3023ff0caef22b2bf4c1b5a18ad1899bd40fa3/frontend/src/plugins/impl/anywidget/AnyWidgetPlugin.tsx#L161-L267
 * Marimo's license is reproduced in LICENSE_MARIMO.
 * @param {function} on_ready Callback when the widget is ready to render (i.e. state is set)
 * @param {function} emit_to_py Send a message to the python backend
 * @returns An AFM object
 */
function createModel(on_ready, emit_to_py, traits) {
  const model = {
    /**
     * Should be true if model.set is triggering callbacks
     * @type {boolean}
     */
    _changing: false,
    /**
     * Changes that have been made, should reset when uploaded to the backend
     * @type {Object<string, any>}
     */
    changed: {},

    /** @type {Comm} */
    _comm: new Comm(null, emit_to_py),

    attributes: null, // Will wait for 'update' message instead.
    callbacks: {},
    get(key) {
      return cleanObject(this.attributes[key]);
    },

    set(key, value) {
      return set_helper(this, key, value);
    },

    set_state(state) {
      // Initial server state.
      if (this.attributes === null) {
        this.attributes = { ...state };
        on_ready(this);
        return;
      }

      this.set(state);
    },

    /** Upload changes to python */
    save_changes() {
      if (this.attributes === null) {
        throw new Error("NiceGUI-Anywidget: save_changes was called before widget was ready (was comm opened?)");
      }

      if (Object.keys(this.changed).length === 0) {
        return;
      }

      // Propagate the change back to python backend;
      // currently serializing all traits instead of just the changed ones
      // (ideally would do this to reduce communication overhead)
      // Could use this.changed to send just a diff.
      // This should NOT run any model.on("change:") callbacks on the frontend.
      this._comm.send_msg('update', { state: this.attributes });

      this.changed = {}
    },
    on(event, callback) {
      if (!this.callbacks[event]) this.callbacks[event] = [];
      this.callbacks[event].push(callback);
    },
    off(event, callback) {
      if (!event) this.callbacks = {};
      else if (!callback) this.callbacks[event] = [];
      else this.callbacks[event]?.delete(callback);
    },
    emit(event, ...values) {
      this.callbacks[event]?.forEach(cb => cb(...values));
      this.callbacks["all"]?.forEach(cb => cb(event));
    },

    send(content, callbacks, buffers) {
      if (callbacks) {
        // It's not clear what the callbacks argument is supposed to do.
        // Marimo seems to pass it to a Promise.then, jupyter takes in a whole object with different functions.
        //  - https://github.com/marimo-team/marimo/blob/7f3023ff0caef22b2bf4c1b5a18ad1899bd40fa3/frontend/src/plugins/impl/anywidget/AnyWidgetPlugin.tsx#L192
        //  - https://github.com/jupyter-widgets/ipywidgets/blob/b24fa6be5a289a23dd82eedcecbf6603ddbe2c0a/packages/base/src/widget.ts#L592
        console.warn('model.send() with callbacks is not supported in NiceGUI currently.');
      }

      this._comm.send_msg('custom', { content }, buffers);
    }
  };
  model._comm.model = model;

  // Serverside updates
  model.on("msg:update", (m, new_state) => {
    m.set_state(new_state);
  });

  return model;
}

export default {
  template: '<div class="nicegui-not-loaded">[NiceGUI-AnyWidget: Waiting for backend connection...]</div>',
  async mounted() {
    const emit_to_py = this.$emit;

    let on_widget_ready;

    // Don't render widget until we have a connection from the server.
    this._waitForComm = new Promise((resolve) => {
      // This should run immediately
      on_widget_ready = resolve;
    }).then(async (m) => {
      let mod;
      try {
        // Dynamically load esm_content as an ECMAScript module
         mod = await load_widget(this.esm_content, this.traits["_anywidget_id"]);
      } finally {
        // We lose the stacktrace if we catch the error.
        // This may happen if the esm has a syntax error, for example.
        this.$el.innerHTML = "[NiceGUI-AnyWidget: Error loading widget (check console for details)]";
        this.$el.classList.add("nicegui-error");
        this.$el.classList.remove("nicegui-not-loaded");
      }

      this.$el.innerHTML = "";
      this.$el.classList.remove("nicegui-not-loaded");

      this.cleanup_widget = await mod.initialize?.({ model: m });
      this.cleanup_view = await mod.render?.({ model: m, el: this.$el });

      this._waitForComm = null;
      on_widget_ready = null;
    });

    if (!on_widget_ready) {
      throw new Error("Promise executor did not run");
    }

    const model = createModel(on_widget_ready, emit_to_py, this.traits);
    this.model = model;

    load_css(this.css_content, this.traits["_anywidget_id"]);
  },
  methods: {
    update_trait(trait, value) {
      this.model.set(trait, value);
    },

    on_delete() {
      this.cleanup_widget?.();
      this.cleanup_view?.();
    },

    update_trait(trait, value) {
        this.model.set(trait, value);
    },

    publish_msg(msg) {
      this.model._comm.recv_msg(msg)
    }
  },
  props: {
    traits: Object,
    esm_content: String,
    css_content: String,
  },
};

File: nicegui/elements/anywidget/anywidget.py
from __future__ import annotations

import importlib.util
import inspect
from pathlib import Path
from typing import TYPE_CHECKING, Any

# Importing this is not ideal. Maybe we could vendor it as well?
from anywidget._util import remove_buffers

from ... import helpers, optional_features
from ...events import GenericEventArguments
from ..mixins.value_element import ValueElement
from . import comm as aw_comm

if importlib.util.find_spec('anywidget'):
    optional_features.register('anywidget')
    if TYPE_CHECKING:
        import anywidget

# Current limitations
# -------------------
# Right now we are manually replacing the comm object in widgets.
# The expected way to work with comms is to import the module, and monkey patch create_comm and create_comm_manager.
#  a) I don't know if we should do this. Nothing else at the moment is using comms so technically it wouldn't interfere, but still.
#     It also means that nicegui *must* be imported before an AnyWidget is constructed.
#  b) The comms should be associated with a NiceGUI element in order to send messages to the frontend, and I'm not too sure how to do that via create_comm.
#     One way I see of doing this is create a comm manager (which should be a singleton) that has a method that allows us to register an element attached to a comm_id.
#     It should probably be in a weakref. But, at the moment, comms work fine so I haven't implemented this.
#
# An immediate disadvantage of replacing the comm manually is that, because they don't go through the comm manager, any registered targets do not run.

class AnyWidget(ValueElement, component='anywidget.js', dependencies=['lib/widget.js', 'lib/set-helper.js'], esm={'nicegui-anywidget': 'dist'}):
    VALUE_PROP: str = 'traits'

    def __init__(self, widget: anywidget.AnyWidget, *, throttle: float = 0) -> None:
        """AnyWidget

        `anywidget <https://anywidget.dev/en/getting-started/>`_ is a library that allows you to
        embed arbitrary JavaScript widgets in a cross-frontend friendly manner.

        There are many publicly available examples of anywidget widgets
        in the `anywidget gallery <https://try.anywidget.dev/>`_, including
        `altair.JupyterChart <https://altair-viz.github.io/user_guide/interactions/jupyter_chart.html>`_,
        and `quak <https://github.com/manzt/quak>`_.

        Implementation: The ``nicegui.anywidget`` element takes an ``AnyWidget`` and observes all ``sync=True`` traits
        of the widget, trigger JS updates when the traits change.
        Conversely, changes on the frontend will be synced back to the widget,
        using ``ValueElement``'s handling to listen to changes on ``traits``.

        *Added in version 3.5.0*

        :param widget: the ``anywidget.AnyWidget`` to wrap
        :param throttle: minimum time (in seconds) between widget updates to Python (default: 0.0)
        """
        self._widget = widget
        self._traits = widget.traits(sync=True)
        super().__init__(value=widget.get_state(self._traits), throttle=throttle)
        self._props['esm_content'] = _get_attribute(widget, '_esm')
        self._props['css_content'] = _get_attribute(widget, '_css')
        self._widget.observe(lambda change: self.run_method('update_trait', change['name'], change['new']), self._traits)

        # We need to replace the widget's comm object in order to communicate with it.
        # More info: https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Low%20Level.html
        # BaseComm.open() is called automatically, which sends messages to the frontend.
        self._widget.comm = aw_comm.create_comm(self, **_get_comm_kwargs(widget))

        # Sent by model.send()
        self.on('anywidget:msg', self._on_widget_msg)

        # Sent by model.save_changes()
        self.on('anywidget:save_changes', self._widget_save_changes)

    def _on_widget_msg(self, content: GenericEventArguments) -> None:
        """Called when comm.send_msg() is called from the frontend."""
        # pylint: disable=protected-access
        self._widget._handle_msg(content.args)

    def _widget_save_changes(self, content: GenericEventArguments) -> None:
        """Called when model.save_changes() is called from the frontend."""
        # ipywidget.set_state handles extra keys in the dict fine.
        # It also uses a context manager to avoid sending events back.
        self._widget.set_state(content.args)

    def on_msg(self, callback, remove=False) -> None:
        """Register the callback with this instance's anywidget.
        Keep in mind that the callback will be called with the AnyWidget, not the NiceGUI element."""
        self._widget.on_msg(callback, remove=remove)

    def _handle_delete(self) -> None:
        self.run_method('on_delete')
        super()._handle_delete()

    def _handle_value_change(self, value: Any) -> None:
        """Update the widget's state when the value changes from frontend"""
        super()._handle_value_change(value)
        state = self._widget.get_state(self._traits)
        for key, value_ in value.items():
            if state[key] != value_:
                setattr(self._widget, key, value_)


def _get_attribute(obj: object, name: str) -> str:
    """Extract the attribute's content, reading if it is a path to a file."""
    content = getattr(obj, name, '')
    if callable(content) and not inspect.isclass(content):  # content is a property function
        content = content()
    assert isinstance(content, (str, Path)), f'Attribute {name} is not a string or Path'
    if helpers.is_file(content):
        content = Path(content).read_text(encoding='utf8')
    assert isinstance(content, str), f'Attribute {name} is a Path but does not exist'
    return content

def _get_comm_kwargs(widget: anywidget.AnyWidget) -> dict:
    """
    Calculates keyword arguments to pass to the comm constructor.

    ipywidgets uses the global comm module functions,
    so this is a manual copy of the widget.open() function.
    """

    # This code is based off of ipywidgets.widgets.widget.Widget.open()
    # https://github.com/jupyter-widgets/ipywidgets/blob/72b939704a6caaef044550a8097388cf934521b4/python/ipywidgets/ipywidgets/widgets/widget.py
    # ipywidget's license is reproduced in LICENSE_JUPYTER.

    state, buffer_paths, buffers = remove_buffers(widget.get_state())
    return {
        'target_name': 'jupyter.widget',
        'data': {'state': state, 'buffer_paths': buffer_paths},
        'buffers': buffers,
        # See here for version information
        # https://github.com/jupyter-widgets/ipywidgets/blob/main/packages/schema/messages.md
        'metadata': {'version': '2.1.0'}
    }
